File: GraphPlan (sgware)/src/edu/uno/ai/planning/gp/TotalOrderPlan.java
package edu.uno.ai.planning.gp;

import java.util.Iterator;

import edu.uno.ai.planning.Plan;
import edu.uno.ai.planning.Step;

public class TotalOrderPlan implements Plan {
	
	public final Step first;
	public final TotalOrderPlan rest;
	public final int size;
	
	private TotalOrderPlan(Step first, TotalOrderPlan rest) {
		this.first = first;
		this.rest = rest;
		this.size = rest.size + 1;
	}
	
	public TotalOrderPlan() {
		this.first = null;
		this.rest = null;
		this.size = 0;
	}

	@Override
	public int size() {
		return size;
	}
	
	private final class MyIterator implements Iterator<Step> {

		private TotalOrderPlan plan = TotalOrderPlan.this;
		
		@Override
		public boolean hasNext() {
			return plan.size > 0;
		}

		@Override
		public Step next() {
			Step next = plan.first;
			plan = plan.rest;
			return next;
		}
	}
	
	@Override
	public Iterator<Step> iterator() {
		return new MyIterator();
	}
	
	public TotalOrderPlan add(Step first) {
		return new TotalOrderPlan(first, this);
	}
}

File: UnitySokoban/Assets/Scripts/Planning/GraphPlanSGW/TotalOrderPlan.cs
using Planning;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Collections;

namespace GraphPlanSGW
{
    public class TotalOrderPlan : Plan
    {
        public readonly Step first;
        public readonly TotalOrderPlan rest;
        public readonly int size;

        private TotalOrderPlan(Step first, TotalOrderPlan rest)
        {
            this.first = first;
            this.rest = rest;
            this.size = rest.size + 1;
        }

        public TotalOrderPlan()
        {
            this.first = null;
            this.rest = null;
            this.size = 0;
        }

        public int Size()
        {
            return size;
        }

        private class MyIterator : IEnumerator<Step>
        {
            private TotalOrderPlan plan;
            private TotalOrderPlan firstPlan;
            private Step current;

            public Step Current { get { return current; } }
            object IEnumerator.Current { get { return current; } }

            public MyIterator(TotalOrderPlan plan)
            {
                this.plan = plan;
                this.firstPlan = plan;
            }

            public void Dispose()
            {
                plan = null;
                firstPlan = null;
                current = null;
            }

            public bool MoveNext()
            {
                if (plan == null)
                    return false;

                current = plan.first;
                plan = plan.rest;

                if (current == null)
                    return false;
                else
                    return true;
            }

            public void Reset()
            {
                plan = firstPlan;
                current = null;
            }
        }

        public TotalOrderPlan add(Step first)
        {
            return new TotalOrderPlan(first, this);
        }

        public IEnumerator<Step> GetEnumerator()
        {
            return new MyIterator(this);
        }

        IEnumerator IEnumerable.GetEnumerator()
        {
            return new MyIterator(this);
        }
    }

}
